Update docs for coco GA release (#365)
NVIDIA/cloud-native-docs: confidential-containers/overview.rst

Reviewer: Minor concern: Do we ever really reference the two illustrations? At least in "Software Components for Confidential Containers" we could refer back to the illustration.

The illustrations are somewhat 'dangling' and not well-explained. Maybe this can be fixed by moving them to where we actually reference these (maybe I was not reading well enough)?

Reviewer: May be good to rebase against https://github.com/NVIDIA/cloud-native-docs/commit/3709f11585c30b2e656858bad286b3dab522eb43 and solve potential conflicts

Reviewer: Is this not duplication of above?

Reviewer: No longer the case with latest bits

Reviewer: This seems outdated, I don't remember what I was referring to

Reviewer: See https://github.com/manuelh-dev/kata-containers/blob/mahuber/doc-update-nvidia-gpu-op/docs/use-cases/NVIDIA-GPU-passthrough-and-Kata-QEMU.md:
`The currently supported modes for enabling GPU workloads in the TEE scenario are: (1) single‑GPU passthrough (one physical GPU per pod) and (2) multi‑GPU passthrough on NVSwitch (NVLink) based HGX systems (for example, HGX Hopper (SXM) and HGX Blackwell / HGX B200).`

Reviewer: No longer the case. We have a new component, the kata lifecycle manager https://github.com/kata-containers/lifecycle-manager - this component is currently in tag v0.1.2 but will still need to be incremented

Reviewer: I don't really know what this means. Does this mean we only support Intel and AMD?

Reviewer: This is no longer the case

Reviewer: General comment: https://github.com/kata-containers/kata-containers/pull/12651/changes has a paragraph/section called "feature set" - to discuss whether we want that here as well, or in the other deployment instructions?

Reviewer 2: ```suggestion
       | Intel EMR / GNR 
```

Reviewer 2: ```suggestion
     - Ubuntu 25.10 
```

Confidential Computing (CC) addresses this gap by using hardware-based Trusted Execution Environments (TEEs), such as AMD SEV-SNP and Intel TDX, with NVIDIA Confidential Computing capabilities to provide isolation, memory encryption, and integrity verification during processing. In addition to isolation, CC provides Remote Attestation, which allows workload owners to cryptographically verify the state of a TEE before providing secrets or sensitive data.

`Confidential Containers <https://github.com/confidential-containers>`__ (CoCo) is the cloud-native approach of CC on Kubernetes.
The Confidential Containers project leverages Kata Containers to provide the sandboxing capabilities. `Kata Containers <https://katacontainers.io/>`_ is an open-source project that provides lightweight Utility Virtual Machines (UVMs) that feel and perform like containers while providing strong workload isolation. Along with the Confidential Containers project, Kata enables the orchestration of secure, GPU-accelerated workloads in Kubernetes.

.. _coco-use-cases:

Use Cases
=========

> The target for Confidential Containers is to enable model providers (Closed and Open source) and Enterprises to use the advancements of Gen AI, agnostic to the deployment model (Cloud, Enterprise, or Edge). Some of the key use cases that CC and Confidential Containers enable are:

Reviewer 3: **Low — Capitalization**: "Closed" and "Open" don't need to be capitalized here since they're not proper nouns.

```suggestion
The target for Confidential Containers is to enable model providers (closed and open source) and Enterprises to use the advancements of Gen AI, agnostic to the deployment model (Cloud, Enterprise, or Edge). Some of the key use cases that CC and Confidential Containers enable are:
```

.. _coco-use-cases:

Use Cases
=========

The target for Confidential Containers is to enable model providers (Closed and Open source) and Enterprises to use the advancements of Gen AI, agnostic to the deployment model (Cloud, Enterprise, or Edge). Some of the key use cases that CC and Confidential Containers enable are:

* **Zero-Trust AI & IP Protection:** You can deploy proprietary models (like LLMs) on third-party or private infrastructure. The model weights remain encrypted and are only decrypted inside the hardware-protected enclave, ensuring absolute IP protection from the host.
* **Data Clean Rooms:** This allows you to process sensitive enterprise data (like financial analytics or healthcare records) securely. Neither the infrastructure provider nor the model builder can see the raw data.

.. image:: graphics/CoCo-Sample-Workflow.png
   :alt: Sample Workflow for Securing Model IP on Untrusted Infrastructure with CoCo

> *Sample Workflow for Securing Model IP on Untrusted Infrastructure with CoCo*

Author: @Hema-Bontha-NV can you share more about the workflow in this diagram. there is the 1-3 steps, but we dont describe them in much detail

NVIDIA's approach to the Confidential Containers architecture delivers on the key promise of Confidential Computing: confidentiality, integrity, and verifiability.
Integrating open source and NVIDIA software components with the Confidential Computing capabilities of NVIDIA GPUs, the Reference Architecture for Confidential Containers is designed to be the secure and trusted deployment model for AI workloads.

The key values of this architecture approach are:

> 1. **Built on OSS standards** - The Reference Architecture for Confidential Containers is built on key OSS components such as Kata, Trustee, QEMU, OVMF, and Node Feature Discovery (NFD), along with hardened NVIDIA components like NVIDIA GPU Operator.

Reviewer 2: ```suggestion
1. **Built on Open Source Software (OSS)** - The Reference Architecture for Confidential Containers is built on key OSS components such as Kata, Trustee, QEMU, OVMF, and Node Feature Discovery (NFD), along with hardened NVIDIA components like NVIDIA GPU Operator.
```

.. list-table::
   :widths: 50 50
   :header-rows: 1

   * - Node A - Traditional Container nodes receive the following software components
     - Node B - Confidential Container nodes receive the following software components
   * - * NVIDIA Driver Manager for Kubernetes
       * NVIDIA Container Toolkit
       * NVIDIA Device Plugin for Kubernetes
       * NVIDIA DCGM and DCGM Exporter
       * NVIDIA MIG Manager for Kubernetes
       * Node Feature Discovery
       * NVIDIA GPU Feature Discovery
     - * NVIDIA Confidential Computing Manager for Kubernetes
       * NVIDIA Sandbox Device Plugin
       * NVIDIA VFIO Manager
       * Node Feature Discovery

This configuration can be controlled through node labelling, as described in the :doc:`Confidential Containers deployment guide <confidential-containers-deploy>`.

Supported Features and Deployment Scenarios
===========================================

The following features are supported with Confidential Containers:

* Support for Confidential Container workloads as

  * Single-GPU passthrough (one physical GPU per pod).
  * Multi-GPU passthrough on NVSwitch (NVLink) based HGX systems.

.. note::

>     For both single and multi GPU Passthrough, all GPUs on the host must be configured for Confidential Computing and all GPUs must be assigned to one Confidential Container virtual machine.

Reviewer 3: **Low — Duplicate content**: This same limitation note ("all GPUs on the host must be configured...") also appears verbatim in the Limitations section at line 213 below. Consider removing one instance to avoid duplication within the same page. (The note also appears in `supported-platforms.rst`, which is fine for cross-page emphasis.)

    Configuring only some GPUs on a node for Confidential Computing is not supported.

* Composite :doc:`attestation <attestation>` using Trustee and the NVIDIA Remote Attestation Service (NRAS).
* Generating Kata Agent Security Policies using the `genpolicy tool <https://github.com/kata-containers/kata-containers/blob/main/src/tools/genpolicy/README.md>`_.
* Use of `signed sealed secrets <https://confidentialcontainers.org/docs/features/sealed-secrets/>`_.
* Access to authenticated registries for container image guest-pull.
* Container image signature verification and encrypted container images.
* Ephemeral container data and image layer storage.
* Lifecycle management of Kata Containers through the `Kata Lifecycle Manager <https://github.com/kata-containers/lifecycle-manager>`_.

More information on these features can be found in the `Confidential Containers documentation <https://confidentialcontainers.org/docs/features/>`_.

Limitations and Restrictions
============================

* NVIDIA supports the GPU Operator and confidential computing with the containerd runtime only.
* Image signature verification for signed multi-arch images is currently not supported.

> * For both single and multi GPU Passthrough, all GPUs on the host must be configured for Confidential Computing and all GPUs must be assigned to one Confidential Container virtual machine.

Reviewer 2: ```suggestion
* For both single and multi GPU Passthrough, all GPUs on the host must be configured for Confidential Computing. For multi GPU Passthrough all GPUs must be assigned to one Confidential Container virtual machine.
```

Author: this should be updated here: https://nvidia.github.io/cloud-native-docs/review/pr-365/confidential-containers/latest/overview.html#limitations-and-restrictions

>   Configuring only some GPUs on a node for Confidential Computing is not supported.

Reviewer 2: ```suggestion
  Configuring only some GPUs on a node for Confidential Computing is not supported.
  
  CoCo provides its own virtualization based on Kata Containers. CoCo cannot run within an existing virtualized environment, i.e., nested virtualization is not supported. CoCo must be installed on the host not within a guest VM.
```

> 

Reviewer 2: ```suggestion

Security Considerations

Application Security Defects: Confidential Computing does not protect against threats within the confidential VM including vulnerabilities in the Application itself. Applications must still follow security best practices such as input validation.

Control Plane: CoCo provides a number of protections to limit the influence of the K8s Control Plane. Applications may expose themselves to attack if they rely upon control plane managed configuration.

Environment Variables and ConfigMaps: Unless provisioned as sealed secrets using Trustee, orchestrator-provided environment variables and ConfigMaps remain untrusted and susceptible to manipulation.

Storage and Volume Mounts: Manipulating volume mounts can expose the workload to malicious untrusted host storage or deny access to required persistent data.

Physical Attacks: Confidential Computing considers most physical attacks out of scope. Your kubernetes hosts should be in physically secure datacenters. 

Availability: Confidential Computing does not provide availability guarantees. Availability is achieved through replication which is a common aspect of k8s deployments.

```

Author: added.

* NVIDIA Confidential Computing Manager (cc-manager) for Kubernetes - to set the confidential computing (CC) mode on the NVIDIA GPUs.

> * NVIDIA Sandbox Device Plugin - to discover NVIDIA GPUs along with their capabilities, to advertise these to Kubernetes, and to allocate GPUs during pod deployment.

Reviewer: https://github.com/kata-containers/kata-containers/pull/12651/changes has an updated version on the description of the sandbox device plugin

> * NVIDIA VFIO Manager - to bind discovered NVIDIA GPUs to the vfio-pci driver for VFIO passthrough.

Reviewer: see https://github.com/kata-containers/kata-containers/pull/12651/changes:
**nvidia-vfio-manager:** Binding discovered NVIDIA GPUs and nvswitches to
  the `vfio-pci` driver for VFIO passthrough.

A minimal hardened init system that securely bootstraps the guest environment, life cycles the kata-agent, provides health checks on started helper daemons while drastically reducing the attack surface.

> Cluster Topology Considerations

Author: This section currently only refers to the gpu operator deployed components. Are there any other deployed componets we should mention here?

Reviewer: There is a kata-deploy pod on the cluster, but not sure if this is worth mentioning here and not sure if this would be worth mentioning at all - asking.

> Cluster Topology Considerations

Author: This section currently only refers to the gpu operator deployed components. Are there any other deployed components we should mention here?

This configuration is done through node labelling and configuration flags set during installation or by editing the ClusterPolicy object post installation. 

Consider the following example where node A is configured to run traditional containers and node B is configured to run confidential containers.

.. list-table::
   :widths: 50 50
   :header-rows: 1

   * - Node A - Traditional Container nodes receive the following software components
     - Node B - Confidential Container nodes receive the following software components
   * - * NVIDIA Driver Manager for Kubernetes
       * NVIDIA Container Toolkit
       * NVIDIA Device Plugin for Kubernetes
       * NVIDIA DCGM and DCGM Exporter
       * NVIDIA MIG Manager for Kubernetes
       * Node Feature Discovery
       * NVIDIA GPU Feature Discovery
     - * NVIDIA Confidential Computing Manager for Kubernetes
       * NVIDIA Sandbox Device Plugin
       * NVIDIA VFIO Manager
       * Node Feature Discovery

This configuration can be controlled through node labelling, as described in the :doc:`Confidential Containers deployment guide <confidential-containers-deploy>`.

> Supported Features and Deployment Scenarios

Author: @Hema-Bontha-NV I've added this section to highlight more supported/available features--does it make sense? any more context you think we should add for folks

        | - NVIDIA VFIO Manager 
        | - NVIDIA Sandbox device plugin 
        | - NVIDIA Confidential Computing Manager for Kubernetes
     - v25.10.0 and higher 

>    * - CoCo release (EA) 

Reviewer: Is this intentional? I think for our latest stack we need a Kata 3.28 release.
I don't know what 'v0.18.0' is here and I ma not sure if we have the exact trustee/guest components in these versions. We are not using a concrete CoCo release. We are using a Kata release and this Kata release pull in CoCo components as dependencies
